python/GafferSceneUITest/HistoryPathTest.py

```diff
@@ -0,0 +1,436 @@
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import unittest
+
+import IECore
+
+import Gaffer
+import GafferTest
+import GafferScene
+import GafferSceneUI
+import GafferSceneTest
+
+class HistoryPathTest( GafferSceneTest.SceneTestCase ) :
+
+	@staticmethod
+	def __inspector( scene, parameter, editScope=None, attribute="light" ) :
+
+		editScopePlug = Gaffer.Plug()
+		editScopePlug.setInput( editScope["enabled"] if editScope is not None else None )
+		inspector = GafferSceneUI.Private.ParameterInspector(
+			scene, editScopePlug, attribute, ( "", parameter )
+		)
+
+		return inspector
+
+	def test( self ) :
+
+		s = Gaffer.ScriptNode()
+
+		s["testLight"] = GafferSceneTest.TestLight()
+
+		s["lightFilter"] = GafferScene.PathFilter()
+		s["lightFilter"]["paths"].setValue( IECore.StringVectorData( [ "/light" ] ) )
+
+		s["tweaks"] = GafferScene.ShaderTweaks()
+		s["tweaks"]["in"].setInput( s["testLight"]["out"] )
+		s["tweaks"]["filter"].setInput( s["lightFilter"]["out"] )
+		s["tweaks"]["shader"].setValue( "light" )
+
+		exposureTweak = Gaffer.TweakPlug( "exposure", 2.0 )
+		s["tweaks"]["tweaks"].addChild( exposureTweak )
+
+		s["editScope"] = Gaffer.EditScope()
+		s["editScope"].setup( s["tweaks"]["out"] )
+		s["editScope"]["in"].setInput( s["tweaks"]["out"] )
+
+		# Just the light
+		inspector = self.__inspector( s["testLight"]["out"], "exposure" )
+
+		with Gaffer.Context() as context :
+			context["scene:path"] = IECore.InternedStringVectorData( ["light"] )
+			historyPath = inspector.historyPath()
+
+		self.assertIsInstance( historyPath, GafferSceneUI.Private.Inspector.HistoryPath )
+
+		self.assertFalse( historyPath.isLeaf() )
+		self.assertTrue( historyPath.isValid() )
+		self.assertEqual( len( historyPath ), 0 )
+
+		c = historyPath.children()
+		self.assertEqual( len( c ), 1 )
+		self.assertTrue( c[0].isLeaf() )
+		self.assertTrue( c[0].isValid() )
+		self.assertEqual( c[0].property( "name" ), "testLight")
+		self.assertEqual( str( c[0] ), "/testLight.out" )
+
+		# Add tweaks
+		inspector = self.__inspector( s["tweaks"]["out"], "exposure" )
+
+		with Gaffer.Context() as context :
+			context["scene:path"] = IECore.InternedStringVectorData( ["light"] )
+			historyPath = inspector.historyPath()
+
+		self.assertFalse( historyPath.isLeaf() )
+		self.assertTrue( historyPath.isValid() )
+		self.assertEqual( len( historyPath ), 0 )
+
+		c = historyPath.children()
+		self.assertEqual( len(c ), 2 )
+		self.assertTrue( c[0].isLeaf() )
+		self.assertTrue( c[0].isValid() )
```

Comment:
We only ever seem to be asserting that a path is valid, and it looks like if we called `path.setFromString( "/gibberish" )` then this path would also claim to be valid. I think we need to override `isValid()`...

Reply:
Fixed in 3ec55b33222cc6fe8e8aad459b50bdda6efbc638